bin/pact-cli.spec.ts

```diff
@@ -0,0 +1,229 @@
+
+			before((done) => server = provider.listen(PORT, () => done()));
+			after(() => server.close());
+
+			it("should work pointing to fake broker", () => {
+				const p = CLI.runSync(["verify", "--provider-base-url", providerBaseUrl, "--pact-urls", path.resolve(__dirname, "integration/me-they-success.json")]).then((cp) => cp.stdout);
+				return expect(p).to.eventually.be.fulfilled;
+			});
+		});
+	});
+});
+
+class CLI {
+	public static runMock(options: ServerOptions = {}): q.Promise<CLI> {
+		const args = _.chain(options)
+			.pairs()
+			.map((arr) => [`--${decamelize(arr[0], "-")}`, `${arr[1]}`])
+			.flatten()
+			.value();
+
+		return this.run(["mock"].concat(args))
+			.tap(() => this.checkMockStarted(options));
+	}
+
+	public static run(args: string[] = []): q.Promise<CLI> {
+		const opts = {
+			cwd: __dirname,
+			detached: !isWindows,
+			windowsVerbatimArguments: isWindows
+		};
+		args = [this.__cliPath].concat(args);
+		const proc = childProcess.spawn("node", args, opts);
+		this.__children.push(proc);
+		return q(new CLI(proc))
+			.tap((cli) => this.commandRunning(cli));
+	}
+
+	public static runSync(args: string[] = []): q.Promise<CLI> {
+		return this.run(args)
+			.tap((cp) => {
+				if ((cp.process as any).exitCode === null) {
+					// console.log("check when exiting");
+					const deferred = q.defer<CLI>();
+					cp.process.once("exit", () => deferred.resolve());
+					return deferred.promise;
+				}
+			});
+	}
+
+	public static remove(proc: ChildProcess) {
+		for (let i = 0; i < this.__children.length; i++) {
+			if (this.__children[i] === proc) {
+				this.__children.splice(i, 1);
+				break;
+			}
+		}
+	}
+
+	public static stopAll() {
+		for (let child of this.__children) {
+			isWindows ? childProcess.execSync(`taskkill /f /t /pid ${child.pid}`) : process.kill(-child.pid, "SIGINT");
+		}
+	}
+
+	private static readonly __children: ChildProcess[] = [];
+	private static readonly __cliPath: string = require.resolve("./pact-cli.js");
+
+	private static commandRunning(c: CLI, amount: number = 0): q.Promise<any> {
+		amount++;
+		const isSet = () => c.stdout.length !== 0 ? q.resolve() : q.reject();
+		return isSet()
+			.catch(() => {
+				if (amount >= 10) {
+					return q.reject(new Error("stdout never set, command probably didn't run"));
+				}
+				return q.delay(1000).then(() => this.commandRunning(c, amount));
+			});
+	}
+
+	private static checkMockStarted(options: ServerOptions, amount: number = 0): q.Promise<any> {
+		amount++;
+		return this.call(options)
+			.catch(() => {
+				if (amount >= 10) {
+					return q.reject(new Error("Pact stop failed; tried calling service 10 times with no result."));
+				}
+				// Try again in 1 second
+				return q.delay(1000).then(() => this.checkMockStarted(options, amount));
+			});
+	}
+
+	private static call(options: ServerOptions): q.Promise<any> {
+		// console.log("Calling to see if pact service is up");
+		options.ssl = options.ssl || false;
```

Comment:
Do we need to do this? Shouldn't TS automatically set them to false/defaults given their explicit type?

Reply:
Yes, we do, because you still need to remember that TS just compiles to javascript, which doesn't have type.  It works fine if someone is using TS with pact-node, but won't work well if someone is using javascript with it.

Reply:
True, it feels like we still have to do a lot of dancing because TypeScript _doesn't_ do this when transpiled back to JS.

Reply:
there is a library that does this for you, but I've yet to try it.

Reply:
Which libraries out of interest?

Reply:
few options here: https://github.com/Microsoft/TypeScript/issues/1573

As well as https://www.npmjs.com/package/runtime-type-checks and https://github.com/gcanti/io-ts

Haven't played with any of them, but would remove the need for any type checking in our own code and let something else handle it all.